src/community/jdbcstore/src/main/java/org/geoserver/jdbcstore/JDBCResourceStore.java

```diff
@@ -0,0 +1,375 @@
+            }
+            try {
+                return new CachingOutputStreamWrapper(File.createTempFile("out.", entry.getName()));
+            } catch (IOException ex) {
+                throw new IllegalStateException(ex); 
+            }
+        }
+
+        @Override
+        public File file() {
+            if (getType() == Type.DIRECTORY) {
+                throw new IllegalStateException("Directory (not a file)");
+            }
+            final Lock lock = lock();
+            try {
+                return cache.cache(this, false);
+            } catch(IOException ex) {
+                throw new IllegalStateException(ex);
+            } finally {
+                lock.release();
+            }
+        }
+
+        @Override
+        public File dir() {
+            if (getType() == Type.RESOURCE) {
+                throw new IllegalStateException("File (not a directory)");
+            }
+            final Lock lock = lock();
+            try {
+                return cache.cache(this, true);
+            } catch(IOException ex) {
+                throw new IllegalStateException(ex);
+            } finally {
+                lock.release();
+            }
+        }
+
+        @Override
+        public long lastmodified() {
+            Timestamp ts = entry.getLastModified();
+            return ts == null ? 0L : ts.getTime();
+        }
+
+        @Override
+        public Resource parent() {
+            return new JDBCResource(entry.getParent());
+        }
+
+        @Override
+        public Resource get(String resourcePath) {
+            return JDBCResourceStore.this.get(path() + "/" + resourcePath);
+        }
+
+        @Override
+        public List<Resource> list() {
+            if (getType() != Type.DIRECTORY) {
+                return Collections.EMPTY_LIST;
+            }            
+
+            List<Resource> list = new ArrayList<Resource>();
+
+            for (JDBCDirectoryStructure.Entry child : entry.getChildren()) {
+                list.add(new JDBCResource(child));
+            }
+
+            return list;
+
+        }
+
+        @Override
+        public Type getType() {
+            Boolean isDir = entry.isDirectory();
+            return isDir == null ? Type.UNDEFINED : isDir ? Type.DIRECTORY : Type.RESOURCE ;
+        }
+
+        @Override
+        public int hashCode() {
+            return entry.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (!(obj instanceof JDBCResource)) {
+                return false;
+            }
+            JDBCResource other = (JDBCResource) obj;
+            return entry.equals(other.entry);
+        }
+
+        @Override
+        public boolean delete() {
+            List<Lock> locks = new ArrayList<Lock>();
```

Comment:
I wonder what is a good idiom for try-with-resources on a collection? Once we have a Closeable Lock, we could have a Closeable LockCollection or Lock CollectionTree (for recursive closure).